README.md
# HaloRay ![](images/hexagon_small.png)

[![Build status](https://ci.appveyor.com/api/projects/status/5k9laekby84x2ex1/branch/develop?svg=true)](https://ci.appveyor.com/project/naavis/haloray/branch/develop)

HaloRay simulates the reflection and refraction of sun light inside hexagonal
ice crystals present in high altitude clouds in the atmosphere. These ice
crystals produce various optical phenomena in the sky, including bright spots,
circles and arcs.

HaloRay employs GPGPU to massively accelerate simulations. The simulation is
done using OpenGL compute and fragment shaders.

HaloRay currently supports Windows and Linux.

An OpenGL 4.4 compliant GPU is required to run HaloRay. On Windows you also need
the [latest Microsoft Visual C++ Redistributable for Visual Studio 2019.](https://aka.ms/vs/16/release/vc_redist.x64.exe)

![Simulation of a column crystal halo display](images/ui-screenshot.png)

## How to use?

HaloRay is designed in such a way that pretty much every simulation parameter
can be tweaked in real time. The simulation traces a given number of light rays
through randomly generated ice crystals per rendered frame, and accumulates the
the results over time. Any time a simulation parameter is changed, the
simulation starts over.

All the simulation parameters are found on the left-hand side of the user
interface. Running the simulation can be started by clicking on the **Render**
button in the bottom left corner.

Camera orientation can be changed by clicking and dragging on the simulated
view, and the mouse scroll wheel can be used to change the field of view.

### General settings

Here are some general settings for the whole simulation.

- **Simulation type**: Choose between parallel-light and divergent-light simulations
  - Parallel-light simulation is suitable for simulating halos in cirrus clouds caused by sun or moon light
  - Divergent-light simulation is suitable for simulating halos in ground-level diamond dust clouds caused by artificial lights
- **Light elevation:** Light source elevation above horizon in degrees
- **Light diameter:** Angular diameter of the light source in degrees
  - This only applies to parallel-light simulations; the light source is always infinitely small for divergent-light simulations
- **Rays per frame:** Number of rays traced through individual crystals per
  rendered frame
  - If the user interface slows down a lot during rendering, lower this value
  - On an NVIDIA GeForce RTX 3070 a good value seems to be around 500 000
  - The maximum value for this parameter may be limited by your GPU
- **Maximum frames:** Simulation stops after rendering this many frames
- **Double scattering:** Probability of a single light ray to scatter from two
  different ice crystals
  - Note that this slows down the simulation significantly!
  - The current implementation only does double scattering inside one crystal
    population
  - A value of 0.0 means no rays are scattered twice, and 1.0 means all rays
    are scattered twice

### Crystal settings

HaloRay allows you to simulate multiple different ice crystal populations
simultaneously. You give each population a name for easier reference by typing
in the **Crystal population** dropdown menu. Each population has a relative
weight, which can be changed by adjusting the **Population weight** slider.
For example, giving weights 1.0 and 3.0 to two crystal populations respectively
would trace three times as many rays through the latter population than the
former. It is also possible to enable or disable a crystal population
temporarily with the **Population enabled** checkbox.

The crystals are hexagonal, and have three named axes as shown in the image
below.

![Graphic of the different crystallographic axes](images/crystal-axes.png)

The orientation of the ice crystals in each population are defined by two
parameters: tilt of the crystal around the A-axis and rotation around the
C-axis. For each parameter you can currently choose between two different
random distributions: a uniform distribution and a Gaussian distribution. For
the Gaussian distribution you can choose an average angle and the standard
deviation of the distribution.

The following table shows parameters needed to simulate crystal orientations
known to happen in nature.

| Orientation | Tilt around a-axis | Rotation around c-axis |
| ----------- | ------------------ | ---------------------- |
| Column      | 90                 | Uniform                |
| Plate       | 0                  | Uniform                |
| Parry       | 90                 | 0                      |
| Lowitz      | Uniform            | 0                      |
| Random      | Uniform            | Uniform                |

The shape of the crystal can also be adjusted by changing the following
parameters:

- **C/A ratio average:** Ratio between the C-axis and A-axis lengths of
  of the crystal
- **C/A ratio std:** Standard deviation of the C/A ratio

![Graphic of plate and column crystals](images/plate-column.png)

Crystals with a large C/A ratio are look like pencil pieces, and are commonly
called column crystals. Column crystals tend to orient themselves with the
C-axis horizontal. Crystals with a small C/A ratio are called plate
crystals. They tend to orient themselves with the C-axis vertical. Both
kinds of crystals are shown in the image above.

The ice crystals also have adjustable pyramidal end caps. Both the upper and
lower caps have the following settings:

- **Apex angle:** Defines the apex angle of the pyramid cap in degrees
  - Ranges from 0.0 to 180.0 degrees, where low numbers mean a very pointy cap
    while high numbers mean the opposite
  - Typically the apex angle is 56 degrees on water ice crystals
- **Apex average height:** Defines the height of the pyramid cap
  - Ranges from 0.0 to 1.0, where 0.0 means the cap is totally flat and 1.0
    means a cap that converges to a sharp point
- **Apex height std:** Standard deviation of the pyramid cap height

![Graphic of pyramidal ice crystal](images/pyramid-crystal.png)

Above is a representation of an ice crystal with pyramidal end caps.
Currently HaloRay is limited to convex ice crystals, so the end caps
cannot extend inwards to make hollow ice crystals.

HaloRay provides six sliders you can use to adjust the **distance of each prism
face** of the hexagonal ice crystals from the crystal C-axis.

![Graphic of a non-regular shaped ice crystal](images/non-regular-crystal.png)

The above image shows a crystal where every other prism face has the default
distance of 1.0 from the C-axis, while every other is reduced to 0.7.

**Note:** Combining non-regular ice crystals with a Parry orientation in
divergent-light simulations might not work as you would expect. The simulation
cannot guarantee that a specific prism face always points in a given direction,
like it can in parallel-light simulations. Any of the prism faces might be
pointing down in the traditional Parry orientation.

### View settings

These settings affect how the results of the simulation are shown on the screen.

- **Camera projection:** Defines how the view is projected to the screen
- **Field of view:** Vertical field of view in degrees
- **Pitch:** Vertical orientation of the camera in degrees from the horizon
- **Yaw:** Horizontal orientation of the camera in degrees from the sun's direction
- **Brightness:** Alters the total brightness of the image, much like an exposure adjustment on cameras
- **Hide sub-horizon:** Hides any halos below the horizon level
- **Lock to light source:** Locks the camera to the sun
- **Show guides:** Draws markings for horizon, zenith, nadir, 22r and 46r circles

### Atmosphere settings

HaloRay renders a realistic sky and sun disk based on a blend of Hosek-Wilkie and
Preetham models. The sky background is only available for parallel-light simulations,
because it does not make much sense in the context of ground-level diamond dust clouds
and street lights.
The sky model has only a few adjustable parameters:

- **Atmosphere enabled:** Toggles rendering of the sky and sun
- **Turbidity:** The amount of aerosols/haze in the atmosphere
- **Ground albedo:** Albedo of the ground plane
  - 0.0 means the ground does not reflect any light
  - 1.0 means the ground reflects all light

### Menus

The top menus should be pretty self-explanatory. Entries in the _File_ menu
allow you to reset the simulation, save and load simulation parameters, and save
the simulation output to an image file on disk.

_View -> Crystal preview_ lets you see a wireframe preview of the an average
ice crystal in the currently selected crystal population.

## How to build?

The user interface is built with [Qt 5](https://www.qt.io/), so you need to
[download the Qt libraries](https://www.qt.io/download-qt-installer) before
compiling HaloRay.

The build is handled with Qt's build tool qmake. You can also use the Qt Creator
IDE to open and build the code.

On Linux you can install Qt using your package manager. On Ubuntu Linux you can
install Qt by running:

```bash
sudo apt-get install qt5-default
```

Finally build the project by running:

```bash
mkdir build
cd build
qmake ../src/haloray.pro
make
```

You can use `nmake` instead of `make` on Windows.

On Windows you need to add the Qt5 binary directory to your PATH environment
variable or copy at least the following Qt DLL files to the same folder as the
resulting executable:

- Qt5Core.dll
- Qt5Widgets.dll
- Qt5Gui.dll
- Qt5Svg.dll

You can also do this automatically with the
[windeployqt](https://doc.qt.io/qt-5/windows-deployment.html) tool, which is
shipped with Qt 5. This is the recommended way.

You can check `scripts\build.ps1` to see how the project is built on the
Appveyor CI server.

## FAQ - Frequently asked questions

### UI components are scaled all wrong on a 4K display in Windows, what to do?

On high-DPI screens Windows scales UI elements by default. HaloRay doesn't fully
adapt to this scaling yet. For now, you can either disable the scaling on an
operating system level by setting UI scale to 100% in Windows display settings,
or try running HaloRay with
[special command-line parameters](https://doc.qt.io/qt-5/qguiapplication.html#platform-specific-arguments)
like this:

```
HaloRay.exe -platform windows:dpiawareness=2
```

### HaloRay crashes or does not start, how to troubleshoot?

HaloRay writes a log file to help in troubleshooting.

On Windows you can find it in `%LOCALAPPDATA%\Temp\haloray\haloray.log` where
`%LOCALAPPDATA%` is usually equal to `C:\Users\<username>\AppData\Local`

On Linux the log file is in `/tmp/haloray/haloray.log`

If the logs do not show any useful information and no error messages pop up,
it is possible the crash happens because of a GPU driver bug. This has
occasionally happened with some Intel GPUs, and there is not much HaloRay
can do about it. It is recommended to update to the latest GPU drivers if
problems like this are encountered.

## Acknowledgments

- [Lauri Kangas](https://github.com/lkangas) for providing tons of reading material and debugging help
- [Panu Lahtinen](https://github.com/pnuu) for additional Linux support
- Jukka Ruoskanen for making HaloPoint 2.0 back in the day and inspiring me to start working on HaloRay
- Marko Riikonen for giving valuable feedback and reporting bugs
- [Jaakko Lehtinen](https://users.aalto.fi/~lehtinj7/) for super valuable lessons in computer graphics


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Added

- Added divergent-light simulation mode to enable simulating halos
  caused by artificial lights in ground-level diamond dust clouds

### Changed

- Changed column and Parry default tilt standard deviation from 1.0
  to 0.5, and same for Parry rotation standard deviation
- All the side panel sections are now expanded by default

## 4.1.0 - 2022-07-23

### Changed

- Changed equation used to calculate index of refraction from one provided in
  Stanley Gedzelman's paper Simulating Rainbows and Halos in Color to a
  polynomial fitted to data provided by Warren and Brandt in their paper Optical
  Constants of Ice from The Ultraviolet to The Microwave, which is what
  HaloPoint 2.0 uses
- HaloRay now remembers which folder you opened last when loading or saving
  files like simulation parameter files or images, and opens the file dialog
  picker in that same folder the next time

## 4.0.1 - 2021-08-26

### Fixed

- Fixed bug where writes outside texture bounds ended up inside the texture
  on some Intel GPUs on Linux
- Fixed guide marking shader to compile on Linux and Intel GPUs

## 4.0.0 - 2021-08-25

### Added

- Option to show 22r and 46r guide circles, horizon line, nadir and zenith

### Changed

- Crystal geometry generation logic was rewritten to keep pyramid apex angle
  constant regardless of varying prism face distances to the crystal C-axis

## 3.3.0 - 2021-05-07

### Changed

- When typing values into spinboxes, the change only takes effect after pressing
  enter or moving focus elsewhere.
- Increased maximum value of image brightness slider

### Fixed

- Fixed bug where typing decimals into spinboxes wasn't working
- Fixed bug where sky was shown a second time outside field of view with
  equidistant and equal area camera projections
- Fixed missing prism face distances from save files
- Fixed bug with slightly incorrect crystal geometry causing uneven brightness
  distribution with Parry crystal halos

## 3.2.0 - 2021-03-24

### Changed

- Improved logging and error handling

### Fixed

- Fixed bug where pyramid apex angle was set incorrectly based on crystal edge
  angle rather than face angle

## 3.1.0 - 2021-02-21

### Changed

- Improved logging and error messages in case shader compilation fails
- Population weight was changed from an integer to a decimal number for more
  granular control

## 3.0.0 - 2021-02-21

### Added

- Created a realistic sky and sun model based on Hosek-Wilkie and Preetham
  papers
- Simulation rate shown in status bar
- Ice crystals now have adjustable pyramidal caps
- Possibility to save and load simulations
- Added controls for prism face distances from the crystal C-axis
- Preview window for crystal shape and orientation
- Better logging and error handling
- Checkbox to enable or disable a crystal population

### Changed

- Forced application to use native desktop OpenGL instead of ANGLE or software
  renderer
- Build system switched to qmake from CMake
- Rewrote most of the GUI code to use Qt model/view architecture
- Triangle normals are now cached during raytracing
- The settings groups in the side panel are now collapsible
- Adjusting sliders with arrow kys now uses smaller steps
- Light is now allowed to bounce inside ice crystal for a 100 times instead
  of 10

### Fixed

- Fixed bug where intensity of halos changed with rays per frame setting
- Fixed a tiny raytracing artifact in the middle of the view

## 2.4.0 - 2019-07-29

### Added

- Show error dialog and exit if GPU does not support OpenGL 4.4

### Fixed

- Bug where changing multiple scattering probability did not trigger a new
  render if maximum iterations were reached
- Crashing on Intel GPUs with Linux and Mesa3D

## 2.3.1 - 2019-07-14

### Fixed

- Build configuration bug where resources were not being linked on Linux

## 2.3.0 - 2019-07-13

### Added

- Added menu bar with option to save simulation result as image
- Added slider for setting the probability of simple double scattering

### Changed

- Replaced Add Population and Remove Population texts with icons

## 2.2.0 - 2019-07-09

### Added

- Created an application icon for HaloRay
- Added possibility to add preset crystal populations

### Changed

- Disabled "Remove population" button when there is only one population in
  simulation
- Added multiple crystal populations by default

## 2.1.1 - 2019-07-08

### Fixed

- Fixed bug where removing the only crystal population caused the UI controls
  not to work anymore

## 2.1.0 - 2019-07-07

### Added

- Gave main window an initial size hint
- Support for multiple crystal populations

### Changed

- Made sidebar scrollable when the window is small enough
- Made UI show field of view in degrees
- Limited field of view for stereographic, rectilinear and orthographic
  projections

### Fixed

- Fixed bug where field of view sometimes did not update correctly when choosing
  a new projection

## 2.0.0 - 2019-07-02

### Changed

- Replaced Nuklear UI with Qt

## 1.1.0 - 2019-06-17

### Changed

- Renamed C-axis orientation to simply C-axis tilt

### Fixed

- Fixed bug where uniform C-axis orientation was messing with C-axis rotation

## 1.0.0 - 2019-06-16

- First official release


src/gui/gui.cpp
#include "gui.h"
#include <functional>
#define NK_INCLUDE_FIXED_TYPES
#define NK_INCLUDE_STANDARD_IO
#define NK_INCLUDE_STANDARD_VARARGS
#define NK_INCLUDE_DEFAULT_ALLOCATOR
#define NK_INCLUDE_VERTEX_BUFFER_OUTPUT
#define NK_INCLUDE_FONT_BAKING
#define NK_INCLUDE_DEFAULT_FONT
#define NK_GLFW_GL3_IMPLEMENTATION
#define NK_KEYSTATE_BASED_INPUT
#include "nuklear/nuklear.h"
#include "../simulation/simulationEngine.h"

namespace GUI
{

const nk_flags WINDOW_FLAGS = NK_WINDOW_BORDER | NK_WINDOW_MOVABLE | NK_WINDOW_SCALABLE | NK_WINDOW_MINIMIZABLE | NK_WINDOW_TITLE;
const nk_flags GROUP_FLAGS = NK_WINDOW_BORDER | NK_WINDOW_TITLE;

void renderGeneralSettingsWindow(struct nk_context *ctx,
                                 bool isRendering,
                                 int &numRays,
                                 const int maxNumRays,
                                 int &maxNumIterations,
                                 const int currentIteration,
                                 HaloSim::LightSource &light,
                                 std::function<void()> renderButtonFn,
                                 std::function<void()> stopButtonFn)
{
    if (nk_begin(ctx, "General settings", nk_rect(50, 20, 330, 430), WINDOW_FLAGS))
    {
        nk_layout_row_dynamic(ctx, 120, 1);
        if (nk_group_begin(ctx, "Sun parameters", GROUP_FLAGS))
        {
            nk_layout_row_dynamic(ctx, 30, 1);
            nk_property_float(ctx, "#Altitude:", -90.0f, &(light.altitude), 90.0f, 0.05f, 0.1f);
            nk_property_float(ctx, "#Diameter:", 0.0f, &(light.diameter), 360.0f, 0.01f, 0.1f);

            nk_group_end(ctx);
        }

        nk_layout_row_dynamic(ctx, 120, 1);
        if (nk_group_begin(ctx, "Simulation parameters", GROUP_FLAGS))
        {
            nk_layout_row_dynamic(ctx, 30, 1);
            nk_property_int(ctx, "#Rays per frame:", 10000, &numRays, maxNumRays, 1000, 5000);
            nk_property_int(ctx, "#Maximum number of frames:", 1, &maxNumIterations, 1000000, 1, 10);

            nk_group_end(ctx);
        }

        nk_layout_row_dynamic(ctx, 50, 1);
        if (isRendering)
        {
            if (nk_button_label(ctx, "Stop"))
            {
                stopButtonFn();
            }
        }
        else
        {
            if (nk_button_label(ctx, "Render"))
            {
                renderButtonFn();
            }
        }
        nk_layout_row_dynamic(ctx, 30, 1);
        nk_label(ctx, "Progress", NK_TEXT_LEFT);
        nk_progress(ctx, (nk_size *)&currentIteration, maxNumIterations, nk_false);
    }
    nk_end(ctx);
}

void renderCrystalSettingsWindow(struct nk_context *ctx,
                                 HaloSim::CrystalPopulation &population)
{
    if (nk_begin(ctx, "Crystal settings", nk_rect(50, 710, 500, 330), WINDOW_FLAGS))
    {
        nk_layout_row_dynamic(ctx, 30, 2);
        nk_property_float(ctx, "#C/A ratio average:", 0.01f, &(population.caRatioAverage), 10.0f, 0.05f, 0.01f);
        nk_property_float(ctx, "#C/A ratio std:", 0.0f, &(population.caRatioStd), 10.0f, 0.05f, 0.01f);

        const char *distributions[] = {"Uniform", "Gaussian"};
        nk_layout_row_dynamic(ctx, 120, 1);
        if (nk_group_begin(ctx, "C axis tilt", GROUP_FLAGS))
        {
            nk_layout_row_dynamic(ctx, 30, 1);
            nk_combobox(ctx, distributions, NK_LEN(distributions), &(population.tiltDistribution), 30, nk_vec2(nk_layout_widget_bounds(ctx).w, 90));
            if (population.tiltDistribution == 1)
            {
                nk_layout_row_dynamic(ctx, 30, 2);
                nk_property_float(ctx, "#Average tilt:", 0.0f, &(population.tiltAverage), 360.0f, 0.1f, 0.5f);
                nk_property_float(ctx, "#Tilt std:", 0.0f, &(population.tiltStd), 360.0f, 0.1f, 0.5f);
            }
            nk_group_end(ctx);
        }

        nk_layout_row_dynamic(ctx, 120, 1);
        if (nk_group_begin(ctx, "Rotation around C axis", GROUP_FLAGS))
        {
            nk_layout_row_dynamic(ctx, 30, 1);
            nk_combobox(ctx, distributions, NK_LEN(distributions), &(population.rotationDistribution), 30, nk_vec2(nk_layout_widget_bounds(ctx).w, 90));
            if (population.rotationDistribution == 1)
            {
                nk_layout_row_dynamic(ctx, 30, 2);
                nk_property_float(ctx, "#Average rotation:", 0.0f, &(population.rotationAverage), 360.0f, 0.1f, 0.5f);
                nk_property_float(ctx, "#Rotation std:", 0.0f, &(population.rotationStd), 360.0f, 0.1f, 0.5f);
            }
            nk_group_end(ctx);
        }
    }
    nk_end(ctx);
}

void renderViewSettingsWindow(struct nk_context *ctx,
                              float &exposure,
                              bool &lockedToSun,
                              HaloSim::Camera &camera,
                              double framesPerSecond)
{
    if (nk_begin(ctx, "View", nk_rect(1550, 700, 330, 320), WINDOW_FLAGS))
    {
        nk_layout_row_dynamic(ctx, 30, 1);
        nk_label(ctx, "Brightness", NK_TEXT_LEFT);
        nk_slider_float(ctx, 0.01f, &(exposure), 10.0f, 0.1f);

        nk_label(ctx, "Field of view", NK_TEXT_LEFT);
        nk_slider_float(ctx, 0.01f, &(camera.fov), 2.0f, 0.01f);

        int locked = lockedToSun ? nk_true : nk_false;
        nk_checkbox_label(ctx, "Lock view to sun", &locked);
        lockedToSun = locked == nk_true;

        int hideSubHorizon = camera.hideSubHorizon ? nk_true : nk_false;
        nk_checkbox_label(ctx, "Hide sub-horizon", &hideSubHorizon);
        camera.hideSubHorizon = hideSubHorizon == nk_true;

        const char *projections[] = {
            "Stereographic",
            "Rectilinear",
            "Equidistant",
            "Equal area",
            "Orthographic"};

        int projectionIndex = (int)camera.projection;
        nk_combobox(ctx, projections, NK_LEN(projections), &(projectionIndex), 30, nk_vec2(nk_layout_widget_bounds(ctx).w, 300));
        camera.projection = (HaloSim::Projection)projectionIndex;

        nk_labelf(ctx, NK_TEXT_LEFT, "Frames per second: %.2f", framesPerSecond);
    }
    nk_end(ctx);
}

} // namespace GUI
